File: bindings/imgui_bundle/demos_cpp/demo_utils/animate_logo.cpp
#define IMGUI_DEFINE_MATH_OPERATORS

#include "demo_utils/animate_logo.h"

#include <map>
#include <fplus/fplus.hpp>
#include "imgui.h"
#include "hello_imgui/hello_imgui.h"
#include "immapp/clock.h"
#include "immapp/browse_to_url.h"


void DrawTransparentImage(ImTextureID texture, ImRect rect, float alpha)
{
    auto alphaColor = ImGui::GetColorU32(ImVec4(1.f, 1.f, 1.f, alpha));
    ImGui::GetForegroundDrawList()->AddImageQuad(
        texture,
        {rect.Min.x, rect.Min.y}, {rect.Max.x, rect.Min.y},
        {rect.Max.x, rect.Max.y}, {rect.Min.x, rect.Max.y},
        ImVec2(0, 0), ImVec2(1, 0), ImVec2(1, 1), ImVec2(0, 1),
        alphaColor
    );
}


void AnimateLogo(const std::string& logoFile, float ratioWidthHeight, ImVec2 emTopRightMargin, float finalAlpha, const char* url)
{
    static double startTime = -1.;
    if (startTime < 0.)
        startTime = ImmApp::ClockSeconds();

    ImTextureID logoTexture = HelloImGui::ImTextureIdFromAsset(logoFile.c_str());

    auto unlerp = [](double a, double b, double x) {
        return (x - a) / (b - a);
    };

    ImRect rect0, rect1;
    float alpha0, alpha1;
    {
        ImVec2 one(1.f, 1.f);
        auto viewportSize = ImGui::GetMainViewport()->Size;
        auto viewportPosition = ImGui::GetMainViewport()->Pos;
        float viewportMinSize = std::min(viewportSize.x, viewportSize.y);

        ImVec2 size0 = ImVec2(viewportMinSize * 0.8f, viewportMinSize * 0.8f / ratioWidthHeight);
        ImVec2 position0 = ImGui::GetMainViewport()->GetCenter() - size0 / 2.f;
        rect0 = ImRect(position0, position0 + size0);
        alpha0 = 1.f;

        float em = ImGui::GetFontSize();
        ImVec2 size1 = ImVec2(viewportMinSize * 0.12f * ratioWidthHeight, viewportMinSize * 0.12f);
        ImVec2 position1 = ImVec2(viewportPosition.x + viewportSize.x - size1.x, viewportPosition.y);
        position1 = position1 + ImVec2(-emTopRightMargin.x * em, emTopRightMargin.y * em);
        rect1 = ImRect(position1, position1 + size1);
        alpha1 = finalAlpha;
    }

    float kAnimation; // between 0 and 1
    {
        double dt = ImmApp::ClockSeconds() - startTime;

        double tPause = 0.4, tAnimation = 0.8;

        if (dt < tPause)
            kAnimation = 0;
        else if (dt < tAnimation)
            kAnimation = unlerp(tPause, tAnimation, dt);
        else
            kAnimation = 1.;
    }

    ImRect rect;
    float alpha;
    {
        rect = ImRect(ImLerp(rect0.Min, rect1.Min, kAnimation), ImLerp(rect0.Max, rect1.Max, kAnimation));
        alpha = ImLerp(alpha0, alpha1, kAnimation);
    }

    {
        if (kAnimation < 1.f)
            HelloImGui::GetRunnerParams()->fpsIdling.enableIdling = false;
        static bool wasIdlingRestored = false;
        if ( (kAnimation >= 1.f ) && ! wasIdlingRestored)
        {
            HelloImGui::GetRunnerParams()->fpsIdling.enableIdling = true;
            wasIdlingRestored = true;
        }
    }

    {
        auto mousePosition = ImGui::GetMousePos();
        if (rect.Contains(mousePosition))
        {
            alpha = 1.f;
            if (ImGui::IsMouseClicked(0))
                ImmApp::BrowseToUrl(url);
        }
    }

    DrawTransparentImage(logoTexture, rect, alpha);
}

File: bindings/imgui_bundle/demos_python/demo_utils/animate_logo.py
import webbrowser
from imgui_bundle import imgui, hello_imgui, immapp, ImVec4, ImVec2

ImTextureID = int


def draw_transparent_image(
    texture: ImTextureID, rect: imgui.internal.ImRect, alpha: float
):
    alpha_color = imgui.get_color_u32(ImVec4(1.0, 1.0, 1.0, alpha))
    imgui.get_foreground_draw_list().add_image_quad(
        texture,
        ImVec2(rect.min.x, rect.min.y),
        ImVec2(rect.max.x, rect.min.y),
        ImVec2(rect.max.x, rect.max.y),
        ImVec2(rect.min.x, rect.max.y),
        ImVec2(0, 0),
        ImVec2(1, 0),
        ImVec2(1, 1),
        ImVec2(0, 1),
        alpha_color,
    )


@immapp.static(start_time=-1.0, was_idling_restored=False)  # type: ignore
def animate_logo(
    logo_file: str,
    ratio_width_height: float,
    em_top_right_margin: ImVec2,
    final_alpha: float,
    url: str,
):
    static = animate_logo
    if static.start_time < 0:
        static.start_time = immapp.clock_seconds()

    logo_texture = hello_imgui.im_texture_id_from_asset(logo_file)

    def unlerp(a, b, x):
        return (x - a) / (b - a)

    def lerp(a, b, x):
        return a + (b - a) * x

    rect0: imgui.internal.ImRect = imgui.internal.ImRect()
    rect1: imgui.internal.ImRect = imgui.internal.ImRect()
    alpha0: float = 1
    alpha1: float = 1

    @immapp.run_anon_block
    def fill_positions():
        nonlocal rect0, rect1, alpha0, alpha1
        ImVec2(1.0, 1.0)
        viewport_size = imgui.get_main_viewport().size
        viewport_position = imgui.get_main_viewport().pos
        viewport_min_size = min(viewport_size.x, viewport_size.y)
        vp_center = imgui.get_main_viewport().get_center()

        size0 = ImVec2(
            viewport_min_size * 0.8, viewport_min_size * 0.8 / ratio_width_height
        )
        position0 = ImVec2(vp_center.x - size0.x / 2, vp_center.y - size0.y / 2)
        rect0 = imgui.internal.ImRect(
            position0, ImVec2(position0.x + size0.x, position0.y + size0.y)
        )
        alpha0 = 1.0

        em = imgui.get_font_size()
        size1 = ImVec2(
            viewport_min_size * 0.12 * ratio_width_height, viewport_min_size * 0.12
        )
        position1 = ImVec2(
            viewport_position.x + viewport_size.x - size1.x, viewport_position.y
        )
        position1 = ImVec2(
            position1.x - em_top_right_margin.x * em,
            position1.y + em_top_right_margin.y * em,
        )
        rect1 = imgui.internal.ImRect(
            position1, ImVec2(position1.x + size1.x, position1.y + size1.y)
        )
        alpha1 = final_alpha

    k_animation: float = 0  # between 0 and 1

    @immapp.run_anon_block
    def fill_k():
        nonlocal k_animation
        dt = immapp.clock_seconds() - static.start_time

        t_pause = 0.4
        t_animation = 0.8

        if dt < t_pause:
            k_animation = 0
        elif dt < t_animation:
            k_animation = unlerp(t_pause, t_animation, dt)
        else:
            k_animation = 1.0

    rect = imgui.internal.ImRect(
        imgui.internal.im_lerp(rect0.min, rect1.min, k_animation),
        imgui.internal.im_lerp(rect0.max, rect1.max, k_animation),
    )
    alpha = lerp(alpha0, alpha1, k_animation)

    if k_animation < 1:
        hello_imgui.get_runner_params().fps_idling.enable_idling = False
    if k_animation >= 1.0 and not static.was_idling_restored:
        hello_imgui.get_runner_params().fps_idling.enable_idling = True
        static.was_idling_restored = True

    mouse_position = imgui.get_mouse_pos()
    if rect.contains(mouse_position):
        alpha = 1
        if imgui.is_mouse_clicked(0):
            webbrowser.open(url)

    draw_transparent_image(logo_texture, rect, alpha)
